onnxruntime/core/framework/execution_frame.cc

```diff
@@ -123,15 +121,36 @@ void IExecutionFrame::Init(const std::vector<int>& feed_mlvalue_idxs, const std:
   }
 
   // 3. handle the weights.
-  // We do this after the fetches to handle an edge case (possibly dubious) where a Constant is an output.
-  // The Constant gets lifted to an initializer so there's no Node producing the value as an output during Graph
-  // execution (i.e. Graph execution won't write the value to all_values_).
+  // We do this after the fetches to handle an edge case where an initializer is an output.
+  // e.g. A Constant node gets lifted to an initializer so there's no Node producing the value as an output during
+  // Graph execution (i.e. Graph execution won't write the value to all_values_).
   // A non-empty fetches vector will overwrite the actual weight in all_values_[ort_value_idx] if we did this earlier.
   // This makes the ONNX Constant test (onnx\backend\test\data\node\test_constant) happy as that
   // involves a graph with a single Constant node.
   for (const auto& entry : initializers) {
     int ort_value_index = entry.first;
-    all_values_[ort_value_index] = entry.second;
+
+    // if the initializer is an output we need to allocate or use a provided fetch buffer and copy the data
+    // so it can be returned to the caller.
+    if (IsOutput(ort_value_index)) {
+      const Tensor& src = entry.second.Get<Tensor>();  // all initializers in ONNX are tensors
```

Comment:
This is in IExecutionFrame::Init whereas AllocateMLValueTensorSelfOwnBuffer is in the derived ExecutionFrame class. We use IExecutionFrame both here and in the optimizers and the latter doesn't have SessionState.

Reply:
Also note ExecutionFrame doesn't trace allocations of graph outputs which is why I kept it to the simpler CopyTensor virtual method.

Reply:
One more thing - the ExecutionFrame alloc methods base their allocations on the allocation plan for the value, but in this case the data is coming from an initializer and not a node output so I believe it would be set to kPreExisting and not the required kAllocateOutput.

Reply:
Clarified with Scott on why not just skip constant folding if it's the graph output instead of this special casing. He'll update code with comments.